device/iot-device-samples/multiplexing-sample/src/main/java/samples/com/microsoft/azure/sdk/iot/ClientManagerBase.java

```diff
@@ -0,0 +1,161 @@
+import static com.microsoft.azure.sdk.iot.device.IotHubConnectionStatusChangeReason.RETRY_EXPIRED;
+import static com.microsoft.azure.sdk.iot.device.transport.IotHubConnectionStatus.DISCONNECTED;
+import static com.microsoft.azure.sdk.iot.device.transport.IotHubConnectionStatus.DISCONNECTED_RETRYING;
+
+@Slf4j
+public abstract class ClientManagerBase implements IotHubConnectionStatusChangeCallback {
+
+    protected enum ConnectionStatus {
+        DISCONNECTED, CONNECTING, CONNECTED
+    }
+
+    protected static final Object lock = new Object();
+    protected static final int SLEEP_TIME_BEFORE_RECONNECTING_IN_SECONDS = 10;
+    protected ConnectionStatus connectionStatus;
+    protected Pair<IotHubConnectionStatusChangeCallback, Object> suppliedConnectionStatusChangeCallback;
+
+    public abstract void openClient() throws IOException;
+    public abstract void closeClient() throws IOException;
+    public abstract String getClientId();
+
+    public void registerConnectionStatusChangeCallback(IotHubConnectionStatusChangeCallback callback, Object callbackContext) {
+        if (callback != null) {
+            this.suppliedConnectionStatusChangeCallback = new Pair<>(callback, callbackContext);
+        } else {
+            this.suppliedConnectionStatusChangeCallback = null;
+        }
+    }
+
+    public void doConnect() throws IOException {
+        // Device client does not have retry on the initial open() call. Will need to be re-opened by the calling application
+        while (connectionStatus == ConnectionStatus.CONNECTING) {
+            synchronized (lock) {
+                if(connectionStatus == ConnectionStatus.CONNECTING) {
+                    try {
+                        log.debug("Opening the device client instance " + getClientId() + " ...");
+                        openClient();
+                        connectionStatus = ConnectionStatus.CONNECTED;
+                        break;
+                    }
+                    catch (Exception ex) {
+                        if (ex.getCause() instanceof TransportException && ((TransportException) ex.getCause()).isRetryable()) {
+                            log.warn("Transport exception thrown while opening DeviceClient instance " + getClientId() + ", retrying: ", ex);
+                        } else {
+                            log.error("Non-retryable exception thrown while opening DeviceClient instance " + getClientId() + ": ", ex);
+                            connectionStatus = ConnectionStatus.DISCONNECTED;
+                            throw ex;
+                        }
+                    }
+                }
+            }
+
+            try {
+                log.debug("Sleeping for 10 secs before attempting another open()");
+                Thread.sleep(SLEEP_TIME_BEFORE_RECONNECTING_IN_SECONDS * 1000);
+            }
+            catch (InterruptedException ex) {
+                throw new RuntimeException("InterruptedException in thread sleep: ", ex);
```

Comment:
I'd just let the method signature declare InterruptedException as a possibility. Any function that waits typically declares it as a possibility. At the very least, the error message here could be something like "Interrupted while waiting between attempting to open the client" to add some extra context